graphql-ballerina/query_depth_validator.bal

```diff
@@ -0,0 +1,103 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import graphql.parser;
+
+class QueryDepthValidator{
+    private parser:DocumentNode documentNode;
+    private int queryDepth = 1;
+    private int maxQueryDepth = 0;
+    private int queryDepthLimit;
+    private ErrorDetail[] errors;
+
+    public isolated function init(parser:DocumentNode documentNode, int queryDepthLimit){
+        self.documentNode = documentNode;
+        self.queryDepthLimit = queryDepthLimit;
+        self.errors = [];
+    }
+
+    public isolated function validate() returns ErrorDetail[]? {
+        self.visitDocument(self.documentNode);
+        if (self.errors.length() > 0) {
+            return self.errors;
+        }
+    }
+
+    public isolated function visitDocument(parser:DocumentNode documentNode) {
+        parser:OperationNode[] operations = documentNode.getOperations();
+        foreach parser:OperationNode operationNode in operations {
+            self.visitOperation(operationNode);
+            if (self.maxQueryDepth > self.queryDepthLimit) {
+                string message = string
+                `Query has depth of ${self.maxQueryDepth}, which exceeds max depth of ${self.queryDepthLimit}`;
+                self.errors.push(getErrorDetailRecord(message, operationNode.getLocation()));
+                return;
+            }
+        }
+    }
+
+    public isolated function visitOperation(parser:OperationNode operationNode) {
+        foreach parser:Selection selection in operationNode.getSelections() {
+            self.visitSelection(selection);
+            if (self.queryDepth > self.maxQueryDepth) {
+                self.maxQueryDepth = self.queryDepth;
+            }
+            self.queryDepth = 0;
+        }
+    }
+
+    public isolated function visitSelection(parser:Selection selection) {
+        if (selection.isFragment) {
+            parser:FragmentNode fragmentNode = self.documentNode.getFragments().get(selection.name);
+            if (fragmentNode.getFields().length() != 0 || fragmentNode.getFragments().length() != 0) {
+                self.visitFragment(fragmentNode);
+            }
+        } else {
+            parser:FieldNode fieldNode = <parser:FieldNode>selection?.node;
+            if (fieldNode.getFields().length() != 0 || fieldNode.getFragments().length() != 0) {
+                self.queryDepth += 1;
+                self.visitField(fieldNode);
+            }
+        }
+    }
+
+    public isolated function visitField(parser:FieldNode fieldNode) {
+        parser:Selection[] selections = fieldNode.getSelections();
+        int i = self.queryDepth;
+        foreach parser:Selection subSelection in selections {
+            self.visitSelection(subSelection);
+            if (self.queryDepth > self.maxQueryDepth) {
+                self.maxQueryDepth = self.queryDepth;
+            }
+            self.queryDepth = i;
```

Comment:
If you increase the query depth when we walk into a `fieldNode`, then we don't need this. We just have to decrement it by `1` when we walk out from the field.